pkg/driver/trace_driver.go

```diff
@@ -236,177 +154,136 @@ func (d *Driver) invokeFunction(metadata *InvocationMetadata) {
+	defer announceFunctionDone.Done()
 
 	function := list.Front().Value.(*common.Function)
-	numberOfInvocations := 0
-	for i := 0; i < len(function.InvocationStats.Invocations); i++ {
-		numberOfInvocations += function.InvocationStats.Invocations[i]
+	invocationCount := getTotalInvocationCount(function.Specification.PerMinuteCount, d.Configuration.TraceDuration, d.Configuration.TraceGranularity)
+	addInvocationsToGroup.Add(invocationCount)
+
+	if invocationCount == 0 {
+		log.Debugf("No invocations found for function %s.\n", function.Name)
+		return
 	}
-	addInvocationsToGroup.Add(numberOfInvocations)
 
-	totalTraceDuration := d.Configuration.TraceDuration
-	minuteIndex, invocationIndex := 0, 0
+	// result statistics
+	minuteIndexSearch := common.NewIntervalSearch(function.Specification.PerMinuteCount)
+	interval := minuteIndexSearch.SearchInterval(0)
+	minuteIndexEnd, minuteIndex, invocationSinceTheBeginningOfMinute := interval.End, interval.Value, 0
 
 	IAT := function.Specification.IAT
+	iatIndex, terminationIAT := 0, invocationCount
 
 	var successfulInvocations int64
 	var failedInvocations int64
-	var failedInvocationByMinute = make([]int64, totalTraceDuration)
 	var numberOfIssuedInvocations int64
 	var currentPhase = common.ExecutionPhase
 
 	waitForInvocations := sync.WaitGroup{}
 
 	if d.Configuration.WithWarmup() {
 		currentPhase = common.WarmupPhase
-		// skip the first minute because of profiling
-		minuteIndex = 1
-
 		log.Infof("Warmup phase has started.")
 	}
 
-	startOfMinute := time.Now()
-	var previousIATSum int64
+	experimentStart := time.Now()
+	startOfIteration := time.Now()
 
 	for {
-		if minuteIndex >= totalTraceDuration {
-			// Check whether the end of trace has been reached
-			break
-		} else if function.InvocationStats.Invocations[minuteIndex] == 0 {
-			// Sleep for a minute if there are no invocations
-			if d.proceedToNextMinute(function, &minuteIndex, &invocationIndex,
-				&startOfMinute, true, &currentPhase, failedInvocationByMinute, &previousIATSum) {
-				break
-			}
-
-			switch d.Configuration.TraceGranularity {
-			case common.MinuteGranularity:
-				time.Sleep(time.Minute)
-			case common.SecondGranularity:
-				time.Sleep(time.Second)
-			default:
-				log.Fatal("Unsupported trace granularity.")
-			}
-
-			continue
+		if iatIndex >= len(IAT) || iatIndex >= terminationIAT {
+			break // end of experiment for this individual function driver
 		}
 
-		iat := time.Duration(IAT[minuteIndex][invocationIndex]) * time.Microsecond
+		iat := time.Duration(IAT[iatIndex]) * time.Microsecond
 
-		currentTime := time.Now()
-		schedulingDelay := currentTime.Sub(startOfMinute).Microseconds() - previousIATSum
-		sleepFor := iat.Microseconds() - schedulingDelay
-		time.Sleep(time.Duration(sleepFor) * time.Microsecond)
+		schedulingDelay := time.Since(startOfIteration)
+		sleepFor := iat - schedulingDelay
+		time.Sleep(sleepFor)
 
-		previousIATSum += iat.Microseconds()
+		startOfIteration = time.Now()
```

Comment:
This would be prone to clock drift over time. I think it is better to have a sum of IAT values from the beginning of the experiment and compare it to the real elapsed time since the beginning of the experiment and correct for the difference. This would correct for the Sleep oversleeping on previous iteration.

Reply:
Going with that approach then.